micro-rdk/src/esp32/certificate.rs

```diff
@@ -28,3 +49,277 @@ impl Certificate for WebRtcCertificate {
+                &mut self.kp_context as *mut mbedtls_pk_context,
+            );
+            mbedtls_x509write_crt_set_version(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                2,
+            );
+            mbedtls_x509write_crt_set_md_alg(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                mbedtls_md_type_t_MBEDTLS_MD_SHA256,
+            );
+        };
+
+        let subject_name = {
+            CString::new(format!(
+                "CN={},O={},C={}",
+                self.oid_cn, self.oid_o, self.oid_c
+            ))
+            .unwrap()
+        };
+        unsafe {
+            MbedTLSError::to_unit_result(mbedtls_x509write_crt_set_issuer_name(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                subject_name.as_ptr(),
+            ))
+        }?;
+
+        unsafe {
+            MbedTLSError::to_unit_result(mbedtls_x509write_crt_set_subject_name(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                subject_name.as_ptr(),
+            ))
+        }?;
+
+        let not_before = CString::new(self.not_before.clone()).unwrap();
+        let not_after = CString::new(self.not_after.clone()).unwrap();
+
+        unsafe {
+            MbedTLSError::to_unit_result(mbedtls_x509write_crt_set_validity(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                not_before.as_ptr(),
+                not_after.as_ptr(),
+            ))
+        }?;
+
+        let mut work_buffer = vec![0_u8; 2048];
+
+        let ret = unsafe {
+            MbedTLSError::to_result(mbedtls_x509write_crt_der(
+                &mut self.crt_context as *mut mbedtls_x509write_cert,
+                work_buffer.as_mut_ptr(),
+                2048,
+                Some(mbedtls_ctr_drbg_random),
+                &mut self.drbg_context as *mut _ as *mut c_void,
+            ))
+        }?;
+        let der_certificate = work_buffer[2048 - ret as usize..].to_vec();
+        assert_eq!(der_certificate.len(), ret as usize);
+
+        unsafe {
+            esp_sha(
+                SHA_TYPE_SHA2_256,
+                der_certificate.as_ptr(),
+                der_certificate.len(),
+                work_buffer.as_mut_ptr(),
+            )
+        };
+        let der_fp = work_buffer
+            .iter()
+            .take(32)
+            .map(|b| format!("{:02X}", b))
+            .collect::<Vec<String>>()
+            .join(":");
+        assert_eq!(der_fp.len(), 95);
+
+        let ret = unsafe {
+            MbedTLSError::to_result(mbedtls_pk_write_key_der(
+                &mut self.kp_context as *mut _,
+                work_buffer.as_mut_ptr(),
+                2048,
+            ))
+        }?;
+
+        let der_kp = work_buffer[2048 - ret as usize..].to_vec();
+        assert_eq!(der_kp.len(), ret as usize);
+
+        let fingerprint = Fingerprint::new("sha-256".to_owned(), der_fp);
```

Comment:
Maybe move this up to line 258 right after you constructed `der_fp`? It seems out of sequence here since what is being built here is `der_kp`.